src/main/java/com/commercetools/sync/services/impl/CustomerServiceImpl.java

```diff
@@ -0,0 +1,128 @@
+
+import com.commercetools.sync.commons.exceptions.SyncException;
+import com.commercetools.sync.customers.CustomerSyncOptions;
+import com.commercetools.sync.services.CustomerService;
+import io.sphere.sdk.commands.UpdateAction;
+import io.sphere.sdk.customers.Customer;
+import io.sphere.sdk.customers.CustomerDraft;
+import io.sphere.sdk.customers.commands.CustomerChangePasswordCommand;
+import io.sphere.sdk.customers.commands.CustomerCreateCommand;
+import io.sphere.sdk.customers.commands.CustomerUpdateCommand;
+import io.sphere.sdk.customers.expansion.CustomerExpansionModel;
+import io.sphere.sdk.customers.queries.CustomerQuery;
+import io.sphere.sdk.customers.queries.CustomerQueryModel;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import static com.commercetools.sync.commons.utils.CtpQueryUtils.buildResourceKeysQueryPredicate;
+import static java.lang.String.format;
+import static java.util.Collections.singleton;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
+
+public final class CustomerServiceImpl extends BaseServiceWithKey<CustomerDraft, Customer, CustomerSyncOptions,
+        CustomerQuery, CustomerQueryModel, CustomerExpansionModel<Customer>> implements CustomerService {
+
+    public CustomerServiceImpl(@Nonnull final CustomerSyncOptions syncOptions) {
+        super(syncOptions);
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Map<String, String>> cacheKeysToIds(
+            @Nonnull final Set<String> keysToCache) {
+        return cacheKeysToIds(keysToCache, keysNotCached -> CustomerQuery
+                        .of()
+                        .withPredicates(buildResourceKeysQueryPredicate(keysNotCached)));
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Set<Customer>> fetchMatchingCustomersByKeys(
+            @Nonnull final Set<String> customerKeys) {
+        return fetchMatchingResources(customerKeys, () -> CustomerQuery
+                        .of()
+                        .withPredicates(buildResourceKeysQueryPredicate(customerKeys)));
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Optional<Customer>> fetchCustomerByKey(@Nullable final String key) {
+        return fetchResource(key, () -> CustomerQuery
+                .of()
+                .withPredicates(buildResourceKeysQueryPredicate(singleton(key))));
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Optional<String>> fetchCachedCustomerId(@Nonnull final String key) {
+        return fetchCachedResourceId(key, () -> CustomerQuery
+                .of()
+                .withPredicates(buildResourceKeysQueryPredicate(singleton(key))));
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Optional<Customer>> createCustomer(
```

Comment:
I guess you need this because the CustomerCreateCommand does not have the right type and can not be changed because it's a different library. But I would suggest here to add a comment on why you have not use the base implementation.

Reply:
@lojzatran cannot reuse it, because `CustomerCreateCommand` replies with `CustomerSignInResult` which does not extend `ResourceView` which is used in `DraftBasedCreateCommand` to resolve the result into a resource object. So, @ahmetoz is right; will add a comment.